File: src/main/java/edu/uchicago/gerber/_08final/mvc/model/Bullet.java
package edu.uchicago.gerber._08final.mvc.model;

import edu.uchicago.gerber._08final.mvc.controller.CommandCenter;
import edu.uchicago.gerber._08final.mvc.controller.Utils;
import lombok.Data;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class Bullet extends Sprite {



    public Bullet(Falcon falcon) {

        setTeam(Team.FRIEND);
        setColor(Color.ORANGE);

        //a bullet expires after 20 frames.
        setExpiry(20);
        setRadius(6);


        //everything is relative to the falcon ship that fired the bullet
        setCenter(falcon.getCenter());

        //set the bullet orientation to the falcon (ship) orientation
        setOrientation(falcon.getOrientation());

        final double FIRE_POWER = 35.0;
        double vectorX =
                Math.cos(Math.toRadians(getOrientation())) * FIRE_POWER;
        double vectorY =
                Math.sin(Math.toRadians(getOrientation())) * FIRE_POWER;

        //fire force: falcon inertia + fire-vector
        setDeltaX(falcon.getDeltaX() + vectorX);
        setDeltaY(falcon.getDeltaY() + vectorY);

        //we have a reference to the falcon passed into the constructor. Let's create some kick-back.
        //fire kick-back on the falcon: inertia - fire-vector / some arbitrary divisor
        final double KICK_BACK_DIVISOR = 36.0;
        falcon.setDeltaX(falcon.getDeltaX() - vectorX / KICK_BACK_DIVISOR);
        falcon.setDeltaY(falcon.getDeltaY() - vectorY / KICK_BACK_DIVISOR);


        //define the points on a cartesian grid
        List<Point> listPoints = new ArrayList<>();
        listPoints.add(new Point(0, 3)); //top point
        listPoints.add(new Point(1, -1)); //right bottom
        listPoints.add(new Point(0, 0));
        listPoints.add(new Point(-1, -1)); //left bottom

        setCartesians(listPoints.toArray(new Point[0]));




    }


    @Override
    public void draw(Graphics g) {
           renderVector(g);
    }
}

File: src/main/java/edu/uchicago/gerber/_08final/mvc/model/Bullet.kt
package edu.uchicago.gerber._08final.mvc.model

import edu.uchicago.gerber._08final.mvc.controller.Utils
import edu.uchicago.gerber._08final.mvc.model.Movable.Team
import java.awt.Color
import java.awt.Graphics
import java.awt.Point
import java.util.*

class Bullet(fal: Falcon) : Sprite() {


    init {
        team = Team.FRIEND
        color = Color.ORANGE

        //a bullet expires after 20 frames. set to one more than frame expiration
        expiry = 21
        radius = 6

        //everything is relative to the falcon ship that fired the bullet
        center = fal.center

        //set the bullet orientation to the falcon (ship) orientation
        orientation = fal.orientation

        val FIRE_POWER = 35.0
        val vectorX = Math.cos(Math.toRadians(orientation.toDouble())) * FIRE_POWER
        val vectorY = Math.sin(Math.toRadians(orientation.toDouble())) * FIRE_POWER

        //fire force: falcon inertia + fire-vector
        deltaX = fal.deltaX + vectorX
        deltaY = fal.deltaY + vectorY


        //fire kick-back on the falcon: inertia - fire-vector / some arbitrary divisor
        val KICK_BACK_DIVISOR = 36.0
        fal.deltaX = fal.deltaX - vectorX / KICK_BACK_DIVISOR
        fal.deltaY = fal.deltaY - vectorY / KICK_BACK_DIVISOR

        //define the points on a cartesian grid
        val listPoint = ArrayList<Point>()
        listPoint.add(Point(0, 3)) //top point
        listPoint.add(Point(1, -1))
        listPoint.add(Point(0, -2))
        listPoint.add(Point(-1, -1))
        cartesians = Utils.pointsListToArray(listPoint)
    }

    override fun draw(g: Graphics) {
        //set the native color of the sprite
        renderVector(g)
    }

}
